contracts/utils/math/Math.sol

```diff
@@ -339,38 +339,69 @@ library Math {
-        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
-        // into the expected uint128 result.
         unchecked {
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            return min(result, a / result);
+            // Take care of easy edge cases when a == 0 or a == 1
+            if (a <= 1) {
+                return a;
+            }
+
+            uint256 aAux = a;
+            uint256 result = 1;
+
+            // For the first guess of `result` (e), we get the biggest power of 2 which is smaller than sqrt(a)
+            // (i.e. 2^e <= sqrt(a)). We know that e is at most 127 given (2^128)^2 overflows an uint256.
+            // Thus, we approximate e by iterating 2^{i/2} where i starts at 128, and applying the exponent
+            // to e if the result is still smaller than a (up to e == 127).
+            if (aAux >= (1 << 128)) {
+                aAux >>= 128;
+                result <<= 64;
+            }
+            if (aAux >= (1 << 64)) {
+                aAux >>= 64;
+                result <<= 32;
+            }
+            if (aAux >= (1 << 32)) {
+                aAux >>= 32;
+                result <<= 16;
+            }
+            if (aAux >= (1 << 16)) {
+                aAux >>= 16;
+                result <<= 8;
+            }
+            if (aAux >= (1 << 8)) {
+                aAux >>= 8;
+                result <<= 4;
+            }
+            if (aAux >= (1 << 4)) {
+                aAux >>= 4;
+                result <<= 2;
+            }
+            if (aAux >= (1 << 2)) {
+                result <<= 1;
+            }
+
+            // We can use the fact that 2^e <= sqrt(a) to improve the estimation
+            // by computing the arithmetic mean between the current estimation and
+            // the next one (result * 2), ensuring that result - sqrt(a) <= 2^{e-2}.
+            result = (3 * result) >> 1;
+
+            // Each Newton iteration will have f(x) = (x + a / x) / 2.
+            // Given the error (ε) is defined by x - sqrt(a), then we know that
+            // ε + 1 == ε^2 / 2x <= ε^2 / 2 * sqrt(a).
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-4.5}
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-9}
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-18}
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-36}
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-72}
+            result = (result + a / result) >> 1; // err := result - sqrt(a) <= 2^{e-144}
```

Comment:
Is that correct?
```suggestion
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 4.5)
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 9)
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 18)
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 36)
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 72)
            result = (result + a / result) >> 1; // ε = ║ result - sqrt(a) ║ < 2**(e - 144)
```
Where does the initial 4.5 come from ? Its not documented anywhere.

Reply:
Yes it's correct. Feel free to update it

Reply:
> Where does the initial 4.5 come from ? Its not documented anywhere.

Added a reference, but comes from using the error formula from Walter Rudin. Principles of Mathematical Analysis. 3rd ed. McGraw-Hill New York, 1976. Exercise 3.16 (b)

Reply:
Rather than adding a reference (to something that I couldn't easily find ... I mean I found a 352 pages long scanned document, with 3.16 being an unrelated definition ... but that doesn't go in the right direction) I'd rather add a proof. 

AFAIK, its faster to reasd the proof than find the reference :P

Reply:
Note 3.16 is not the exercise but a section. I think you're looking for this:

<img width="813" alt="Captura de pantalla 2024-02-15 a la(s) 12 45 15 p m" src="https://github.com/OpenZeppelin/openzeppelin-contracts/assets/33379285/ea038528-4bbd-42bd-9b46-b064216d85b7">

And the solution:

<img width="818" alt="Captura de pantalla 2024-02-15 a la(s) 12 45 29 p m" src="https://github.com/OpenZeppelin/openzeppelin-contracts/assets/33379285/ae7be78d-48cf-4be7-8ff8-fc528734f857">